chore: (bc) Update PHP and PHPUnit minimum version requirements

Pixelshaped opened the pull request:
- Use attributes to handle what was previously handled through annotations

Fixes #163

Pixelshaped commented:
@ilario-pierbattista All tests are passing but 1: I skipped `testReproducibilityWithSeed` for now and will work towards a solution.

Pixelshaped commented:
Found a way to print ERIS_SEED to console. 
But it seems it will be hard to test, see https://github.com/sebastianbergmann/phpunit/issues/4953

ilario-pierbattista commented on composer.json:
```diff
@@ -17,10 +17,10 @@
     "license": "MIT",
     "require": {
-        "php": ">=7.1"
+        "php": ">=8.1"
```
PHP 8.0 has still security fixed until end 2023

Pixelshaped replied:
To be fair, I hadn't PHP 8.0 installed on my machine so I went the quickest way. But I think you're right. Even if it's a breaking change we should maximize compatibility. I'll do it.

Pixelshaped replied:
Oh but wait. PHPUnit 10 has a minimum requirement of PHP 8.1. So I can't go lower and maintain compatibility.

ilario-pierbattista replied:
```suggestion
        "php": "^8.1"
```

ilario-pierbattista reviewed (commented):
Hi @Pixelshaped shouln't we continue to support PHP 8.0?

ilario-pierbattista commented:
Also, this may resolve #160

Pixelshaped commented:
As said in comment, we can't target PHP 8.0 if we're to target PHPUnit10. PHP 8.1 is the minimum. Maybe when publishing it would be nice to change the major version number of the semver notation, so the BC won't surprise people. @ilario-pierbattista

Pixelshaped commented:
@ilario-pierbattista can you tell me what's the best way moving forward? I'd rather avoid publishing my fork. Can't we just advance the major version number so that people aren't trapped into updating?

ilario-pierbattista commented:
Hi @Pixelshaped just changed the base branch to `release-1.x`. 
I can create a rc for the new major, and you can start using it.

However, some extra work will be required to make it stable.

Pixelshaped commented:
I fixed the minimum requirement for PHPUnit to be 10.0.4 (all versions between 10.0 and 10.0.4 triggered an additional issue in a test)
I updated PHP-CS-Fixer and ran it and commited the result.

ilario-pierbattista commented:
This is stil failing https://github.com/giorgiosironi/eris/actions/runs/4762687870/jobs/8470026118?pr=164 any idea?

Pixelshaped commented:
> This is stil failing https://github.com/giorgiosironi/eris/actions/runs/4762687870/jobs/8470026118?pr=164 any idea?

Not sure why, I'll look into it this weekend or next week. :)

ilario-pierbattista commented:
@Pixelshaped Merged `master` in your fork after releasing this https://github.com/giorgiosironi/eris/releases/tag/0.14.0 .

I hope I didn't break anything resolving conflicts.

Pixelshaped commented:
Hi @ilario-pierbattista! I fixed some static tests for Psalm and PHPStan. All scripts defined in composer now produce 0 error (except phpstan-baseline which isn't used by the CI). I'm hoping we're in the clear!

Pixelshaped commented:
I missed `./phpunit --testsuite EndToEnd`... on it.

ilario-pierbattista commented on composer.json:
```diff
@@ -17,10 +17,10 @@
     },
     "require-dev": {
-        "phpunit/phpunit": "^8 || ^9",
+        "phpunit/phpunit": ">=10.0.4",
```
```suggestion
        "phpunit/phpunit": "^10.0.4",
```

Pixelshaped commented:
We still have an issue with `./phpunit --testsuite EndToEnd` which produces a warning `Class PHPUnitDeprecationHelper cannot be found in /Users/renaud/Code/eris/test/PHPUnitDeprecationHelper.php` that I'm still trying to figure out. Probably a namespace issue but not sure yet. As PHPUnit10 is the only tolerated version, I'm inclined to remove `PHPUnitDeprecationHelper` altogether. What do you think?

Pixelshaped commented:
@ilario-pierbattista I decided to remove `PHPUnitDeprecationHelper`. All the fallback methods it included were in any case not possible to reach due to PHPUnit semver requirements.

ilario-pierbattista reviewed (approved):
LGTM

ilario-pierbattista commented:
@Pixelshaped thank you, just released as 1.0.0-rc1 https://github.com/giorgiosironi/eris/releases/tag/1.0.0-rc1
